target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4018-a42.dts

```diff
@@ -169,7 +169,7 @@
 	pinctrl-0 = <&spi_0_pins>;
 	pinctrl-names = "default";
 	status = "okay";
-	cs-gpios = <&tlmm 54 0>;
+	cs-gpios = <&tlmm 54 GPIO_ACTIVE_HIGH>;
```

Comment:
https://github.com/openwrt/openwrt/blob/master/target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4018-a42.dts#L180

The pin_cs node sets
output-high;

If this is used to invert the gpio's polarity, wouldn't it be better to delete the pin_cs output-high property and just set the cs-gpios to GPIO_ACTIVE_LOW?

Reply:
I already wanted to answer to this yesterday because I didn't get your point. But I've decided to sleep on it and hope that I get it today. Unfortunately, this didn't help :)

Let me try to explain it from my perspective:

* `output-high` in pin_cs is there to switch the gpio from in to output direction
   - high is only used to give it a defined initial state
   - low would also have "worked" in some sense because the spi driver changes it to whatever it wants anyway
* cs-gpios[0] with GPIO_ACTIVE_LOW should in theory not work correctly by itself because the slave would not have been selected correctly
  - spi-cs-high property would have been required to work around the wrong GPIO_ACTIVE_LOW (but why use a workaround when you can just set it correctly to GPIO_ACTIVE_HIGH)
  - in reality, something in the spi subsystem (or something else here) is just ignoring the parameter for polarity and only the spi-cs-high is applied in spi_set_cs

Let me just show it a couple of examples.

1.  current state:

        root@OpenWrt:/# grep gpio54 /sys/kernel/debug/gpio 
         gpio54  : out 0 2mA no pull
        root@OpenWrt:/#  dmesg|grep -e spi0.0 |grep -v 'Kernel'
        [    0.551122] m25p80 spi0.0: w25q256 (32768 Kbytes)
        [    0.551187] 10 cmdlinepart partitions found on MTD device spi0.0
        [    0.554832] Creating 10 MTD partitions on "spi0.0":

2. output-high replaced with output-low:

        root@OpenWrt:/# grep gpio54 /sys/kernel/debug/gpio 
         gpio54  : out 0 2mA no pull
        root@OpenWrt:/# dmesg|grep -e spi0.0 |grep -v 'Kernel'
        [    0.550952] m25p80 spi0.0: w25q256 (32768 Kbytes)
        [    0.551018] 10 cmdlinepart partitions found on MTD device spi0.0
        [    0.554661] Creating 10 MTD partitions on "spi0.0"


3.  with cs-gpios[0] set to GPIO_ACTIVE_LOW and output-high removed
    
        root@OpenWrt:/# grep gpio54 /sys/kernel/debug/gpio 
         gpio54  : in  0 2mA no pull
        root@OpenWrt:/# dmesg|grep -e spi0.0 |grep -v 'Kernel'
        [    0.541523] m25p80 spi0.0: unrecognized JEDEC id bytes: 00, 00, 00
        [    0.548069] m25p80: probe of spi0.0 failed with error -2

4. spi-cs-high set for m25p80@0

        root@OpenWrt:/# grep gpio54 /sys/kernel/debug/gpio 
         gpio54  : out 0 2mA no pull
        root@OpenWrt:/# dmesg|grep -e spi0.0 |grep -v 'Kernel'
        [    0.541811] m25p80 spi0.0: unrecognized JEDEC id bytes: 00, 00, 00
        [    0.548391] m25p80: probe of spi0.0 failed with error -2

Maybe it is just a misunderstanding from my side.

Reply:
> spi-cs-high property would have been required to work around the wrong GPIO_ACTIVE_LOW (but why use a workaround when you can just set it correctly to GPIO_ACTIVE_HIGH)
> in reality, something in the spi subsystem (or something else here) is just ignoring the parameter for polarity and only the spi-cs-high is applied in spi_set_cs

Looks like you are right about your assumption there. 
of_spi_register_master() only allocates a single int for the gpio pin/number:
https://github.com/torvalds/linux/blob/master/drivers/spi/spi.c#L2038
it only extracts the gpio number/pin from the property:
https://github.com/torvalds/linux/blob/master/drivers/spi/spi.c#L2049

And likewise spi_set_cs() only takes the spi-cs-high into account.

    static void spi_set_cs(struct spi_device *spi, bool enable)
    {
	if (spi->mode & SPI_CS_HIGH)
		enable = !enable;

	if (gpio_is_valid(spi->cs_gpio)) {
		gpio_set_value(spi->cs_gpio, !enable);

so yeah, the gpio flags in the cs-gpios property are ignored.
https://github.com/torvalds/linux/blob/master/drivers/spi/spi.c#L726

>   2. output-high replaced with output-low:
>    
>     root@OpenWrt:/# grep gpio54 /sys/kernel/debug/gpio 
>      gpio54  : out 0 2mA no pull
>     root@OpenWrt:/# dmesg|grep -e spi0.0 |grep -v 'Kernel'
>     [    0.550952] m25p80 spi0.0: w25q256 (32768 Kbytes)
>     [    0.551018] 10 cmdlinepart partitions found on MTD device spi0.0
>     [    0.554661] Creating 10 MTD partitions on "spi0.0"
heh. ok so it's all about setting the gpio/pin to output then.

> 3. with cs-gpios[0] set to GPIO_ACTIVE_LOW and output-high removed
>
>    root@OpenWrt:/# grep gpio54 /sys/kernel/debug/gpio 
>      gpio54  : in  0 2mA no pull
>    root@OpenWrt:/# dmesg|grep -e spi0.0 |grep -v 'Kernel'
>    [    0.541523] m25p80 spi0.0: unrecognized JEDEC id bytes: 00, 00, 00
>    [    0.548069] m25p80: probe of spi0.0 failed with error -2
>
Congratulations. I think you found a bona fide meta bug in the spi code there. 
it never set the gpio to output... which is definitely should since the
spi_set_cs() function does use gpio_set_value(). And gpio_set_value
assumes that the gpio is already set to output as per <https://lwn.net/Articles/532714/>.

the only saving grace I could find was this comment from 2008 in the vicinity:
https://github.com/torvalds/linux/blob/master/drivers/spi/spi.c#L540

	if (ctlr->cs_gpios)
		spi->cs_gpio = ctlr->cs_gpios[spi->chip_select];

	/* Drivers may modify this initial i/o setup, but will
	 * normally rely on the device being setup.  Devices
	 * using SPI_CS_HIGH can't coexist well otherwise...
	 */
	status = spi_setup(spi);

But yeah, I think the spi code should be fixed.

That said, I don't know what's better in this case now? This is could be very ugly and
simply doing any changes to this code could break existing setups (that used the wrong
values/settings). @CyrillePitchen